Exit code = 0, когда утилита падает с ошибкой, должно быть не 0.

```
$ codeplag -ext py -d
usage: CODEPLAG [-h] [--mode {many_to_many}] [-t {50, 51, ..., 99}] [-v] -ext {py,cpp} [-ab] [-e REGEXP] [-gf GITHUB_FILE [GITHUB_FILE ...]]
                [-g GITHUB_USER] [-gp GITHUB_PROJECT_FOLDER [GITHUB_PROJECT_FOLDER ...]] [-f FILE [FILE ...]] [-d DIRECTORY [DIRECTORY ...]]
CODEPLAG: error: argument -d/--directories: expected at least one argument
$ echo $?
0


$ codeplag -ext py -d jksljdf
Mar 12 16:54:03 - [ERROR] - Directory jksljdf not found or not a directory
$ echo $?
0

$ codeplag -gf https://github.com/jsldjf/jsfkdjf/lksdjfkj/sfd/sdf -ext py
Mar 12 16:56:12 - [INFO] - Getting GitHub files from urls
Mar 12 16:56:12 - [ERROR] - GitHub API rate limit exceeded for 188.243.183.247. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)
$ echo $?
0
```

--- a/.github/workflows/build_and_test.yml
+++ b/.github/workflows/build_and_test.yml
@@ -47,13 +47,4 @@ jobs:
       - name: Run autotests
         run: |
           cd ~/work/code-plagiarism/code-plagiarism
-          codeplag --version && \
-          codeplag --directories src/codeplag/cplag/tests/data src/ --files src/codeplag/cplag/tests/data/sample1.cpp src/codeplag/cplag/tests/data/sample2.cpp -ext cpp && \
-          codeplag -ext py --files ./src/codeplag/pyplag/astwalkers.py --directories ./src/codeplag/pyplag && \
-          codeplag -ext py --directories ./src/codeplag/algorithms ./src && \
-          codeplag -ext py --files src/codeplag/pyplag/astwalkers.py --github-user OSLL --regexp "code-" --all-branches && \
-          codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py --github-user OSLL -e "code-" -ab && \
-          codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py -d src/codeplag/pyplag/ && \
-          codeplag -ext py --directories src/ --github-user OSLL -e "code-" && \
-          codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --github-user OSLL -e "code-" && \
-          codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --directories src/codeplag/pyplag/
+          ./autotests

--- a/.github/workflows/build_and_test.yml
+++ b/.github/workflows/build_and_test.yml
@@ -25,3 +25,35 @@ jobs:
       - name: Run tests in the docker container based on the created image
         run: |
           make docker-test
+
+  autotests:
+    runs-on: ubuntu-18.04
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Install codeplag
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y python3-pip
+          sudo apt-get install -y clang libncurses5
+          sudo apt-get install -y man
+          sudo pip3 install --upgrade setuptools
+          sudo pip3 install argparse-manpage
+          cd /usr/src/
+          sudo git clone https://github.com/OSLL/code-plagiarism.git
+          cd code-plagiarism/
+          sudo make install
+
+      - name: Run autotests
+        run: |
+          cd ~/work/code-plagiarism/code-plagiarism
+          codeplag --version && \
+          codeplag --directories src/codeplag/cplag/tests/data src/ --files src/codeplag/cplag/tests/data/sample1.cpp src/codeplag/cplag/tests/data/sample2.cpp -ext cpp && \
+          codeplag -ext py --files ./src/codeplag/pyplag/astwalkers.py --directories ./src/codeplag/pyplag && \
+          codeplag -ext py --directories ./src/codeplag/algorithms ./src && \
+          codeplag -ext py --files src/codeplag/pyplag/astwalkers.py --github-user OSLL --regexp "code-" --all-branches && \
+          codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py --github-user OSLL -e "code-" -ab && \
+          codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py -d src/codeplag/pyplag/ && \
+          codeplag -ext py --directories src/ --github-user OSLL -e "code-" && \
+          codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --github-user OSLL -e "code-" && \
+          codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --directories src/codeplag/pyplag/

--- a/.github/workflows/build_and_test.yml
+++ b/.github/workflows/build_and_test.yml
@@ -47,4 +47,4 @@ jobs:
       - name: Run autotests
         run: |
           cd ~/work/code-plagiarism/code-plagiarism
-          ./autotests.sh
+          sudo ./autotests.sh

--- a/.github/workflows/build_and_test.yml
+++ b/.github/workflows/build_and_test.yml
@@ -47,4 +47,4 @@ jobs:
       - name: Run autotests
         run: |
           cd ~/work/code-plagiarism/code-plagiarism
-          sudo ./autotests.sh
+          sh ./autotests.sh

--- a/.github/workflows/build_and_test.yml
+++ b/.github/workflows/build_and_test.yml
@@ -47,4 +47,4 @@ jobs:
       - name: Run autotests
         run: |
           cd ~/work/code-plagiarism/code-plagiarism
-          ./autotests
+          ./autotests.sh

--- a/.github/workflows/build_and_test.yml
+++ b/.github/workflows/build_and_test.yml
@@ -45,6 +45,9 @@ jobs:
           sudo make install
 
       - name: Run autotests
+        env:
+            ACCESS_TOKEN: ${{ secrets.ACCESS_TOKEN }}
         run: |
           cd ~/work/code-plagiarism/code-plagiarism
+          echo "ACCESS_TOKEN = '${ACCESS_TOKEN}'" > .env
           sh ./autotests.sh

--- a/autotests.sh
+++ b/autotests.sh
@@ -4,9 +4,11 @@ codeplag --version && \
 codeplag --directories src/codeplag/cplag/tests/data src/ --files src/codeplag/cplag/tests/data/sample1.cpp src/codeplag/cplag/tests/data/sample2.cpp -ext cpp && \
 codeplag -ext py --files ./src/codeplag/pyplag/astwalkers.py --directories ./src/codeplag/pyplag && \
 codeplag -ext py --directories ./src/codeplag/algorithms ./src && \
-codeplag -ext py --files src/codeplag/pyplag/astwalkers.py --github-user OSLL --regexp "code-" --all-branches && \
+(codeplag -ext py --files src/codeplag/pyplag/astwalkers.py --github-user OSLL --regexp "code-" --all-branches || exit $?) && \
 codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py --github-user OSLL -e "code-" -ab && \
 codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py -d src/codeplag/pyplag/ && \
 codeplag -ext py --directories src/ --github-user OSLL -e "code-" && \
 codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --github-user OSLL -e "code-" && \
 codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --directories src/codeplag/pyplag/
+
+echo $?

--- a/autotests.sh
+++ b/autotests.sh
@@ -1,0 +1,12 @@
+#!/bin/bash
+
+codeplag --version && \
+codeplag --directories src/codeplag/cplag/tests/data src/ --files src/codeplag/cplag/tests/data/sample1.cpp src/codeplag/cplag/tests/data/sample2.cpp -ext cpp && \
+codeplag -ext py --files ./src/codeplag/pyplag/astwalkers.py --directories ./src/codeplag/pyplag && \
+codeplag -ext py --directories ./src/codeplag/algorithms ./src && \
+codeplag -ext py --files src/codeplag/pyplag/astwalkers.py --github-user OSLL --regexp "code-" --all-branches && \
+codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py --github-user OSLL -e "code-" -ab && \
+codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py -d src/codeplag/pyplag/ && \
+codeplag -ext py --directories src/ --github-user OSLL -e "code-" && \
+codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --github-user OSLL -e "code-" && \
+codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --directories src/codeplag/pyplag/

--- a/autotests.sh
+++ b/autotests.sh
@@ -4,11 +4,9 @@ codeplag --version && \
 codeplag --directories src/codeplag/cplag/tests/data src/ --files src/codeplag/cplag/tests/data/sample1.cpp src/codeplag/cplag/tests/data/sample2.cpp -ext cpp && \
 codeplag -ext py --files ./src/codeplag/pyplag/astwalkers.py --directories ./src/codeplag/pyplag && \
 codeplag -ext py --directories ./src/codeplag/algorithms ./src && \
-(codeplag -ext py --files src/codeplag/pyplag/astwalkers.py --github-user OSLL --regexp "code-" --all-branches || exit $?) && \
+codeplag -ext py --files src/codeplag/pyplag/astwalkers.py --github-user OSLL --regexp "code-" --all-branches && \
 codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py --github-user OSLL -e "code-" -ab && \
 codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py -d src/codeplag/pyplag/ && \
 codeplag -ext py --directories src/ --github-user OSLL -e "code-" && \
 codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --github-user OSLL -e "code-" && \
 codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --directories src/codeplag/pyplag/
-
-echo $?

--- a/autotests.sh
+++ b/autotests.sh
@@ -1,12 +1,9 @@
 #!/bin/bash
 
 codeplag --version && \
-codeplag --directories src/codeplag/cplag/tests/data src/ --files src/codeplag/cplag/tests/data/sample1.cpp src/codeplag/cplag/tests/data/sample2.cpp -ext cpp && \
-codeplag -ext py --files ./src/codeplag/pyplag/astwalkers.py --directories ./src/codeplag/pyplag && \
-codeplag -ext py --directories ./src/codeplag/algorithms ./src && \
-codeplag -ext py --files src/codeplag/pyplag/astwalkers.py --github-user OSLL --regexp "code-" --all-branches && \
-codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py --github-user OSLL -e "code-" -ab && \
-codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py -d src/codeplag/pyplag/ && \
-codeplag -ext py --directories src/ --github-user OSLL -e "code-" && \
-codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --github-user OSLL -e "code-" && \
-codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag --directories src/codeplag/pyplag/
+codeplag -ext cpp --files src/codeplag/cplag/tests/data/sample1.cpp src/codeplag/cplag/tests/data/sample2.cpp && \
+codeplag -ext cpp --directories src/codeplag/cplag/tests/data && \
+codeplag -ext py --directories src/codeplag/cplag/tests/data src/codeplag/cplag/tests --files ./src/codeplag/pyplag/astwalkers.py && \
+codeplag -ext py --github-files https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/utils.py https://github.com/OSLL/code-plagiarism/blob/main/setup.py && \
+codeplag -ext py --github-user OSLL --regexp "code-plag" && \
+codeplag -ext py --github-project-folders https://github.com/OSLL/code-plagiarism/blob/main/src/codeplag/pyplag/tests

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,3 +1,4 @@
+argcomplete==2.0.0
 numpy==1.19.4
 pandas==1.0.5
 ccsyspath==1.1.0

--- a/src/codeplag/__main__.py
+++ b/src/codeplag/__main__.py
@@ -30,22 +30,20 @@ logger = get_logger(__name__, LOG_PATH)
 
 if __name__ == '__main__':
     logger.debug("Starting codeplag util")
-    try:
-        env_config = Config(RepositoryEnv('./.env'))
-    except FileNotFoundError:
-        ACCESS_TOKEN = ''
-        logger.debug('The environment file is not defined')
-    else:
-        ACCESS_TOKEN = env_config.get('ACCESS_TOKEN', default='')
-        if not ACCESS_TOKEN:
-            logger.debug('GitHub access token is not defined')
 
     parser = get_parser()
     args = vars(parser.parse_args())
 
-    logger.debug("Mode: {}; Extension: {}".format(args['mode'], args['extension']))
+    logger.debug(
+        "Mode: {}; Extension: {}; Environment path: {}.".format(
+            args['mode'],
+            args['extension'],
+            args['environment']
+        )
+    )
 
     MODE = args.pop('mode')
+    ENVIRONMENT = args.pop('environment')
     EXTENSION = args.pop('extension')
     THRESHOLD = args.pop('threshold')
     BRANCH_POLICY = args.pop('all_branches')
@@ -55,6 +53,14 @@ if __name__ == '__main__':
     GITHUB_FILES = args.pop('github_files')
     GITHUB_PROJECT_FOLDERS = args.pop('github_project_folders')
     GITHUB_USER = args.pop('github_user')
+
+    if ENVIRONMENT:
+        env_config = Config(RepositoryEnv(ENVIRONMENT))
+        ACCESS_TOKEN = env_config.get('ACCESS_TOKEN', default='')
+        if not ACCESS_TOKEN:
+            logger.debug('GitHub access token is not defined.')
+    else:
+        ACCESS_TOKEN = ''
 
     begin_time = perf_counter()
     if MODE == 'many_to_many':

--- a/src/codeplag/brand_consts.py
+++ b/src/codeplag/brand_consts.py
@@ -1,7 +1,7 @@
 import argparse
 
 
-UTIL_VERSION = "0.0.7"
+UTIL_VERSION = "0.0.8"
 UTIL_NAME = "codeplag"
 
 

--- a/src/codeplag/codeplagcli.py
+++ b/src/codeplag/codeplagcli.py
@@ -12,7 +12,7 @@ logger = get_logger(__name__, LOG_PATH)
 
 def dir_path(path_to_directory):
     if not os.path.isdir(path_to_directory):
-        logger.error("Directory {} not found or not a directory".format(path_to_directory))
+        logger.error("Directory {} not found or not a directory.".format(path_to_directory))
         exit(1)
 
     return path_to_directory
@@ -20,10 +20,18 @@ def dir_path(path_to_directory):
 
 def file_path(path_to_file):
     if not os.path.isfile(path_to_file):
-        logger.error("File {} not found or not a file".format(path_to_file))
+        logger.error("File {} not found or not a file.".format(path_to_file))
         exit(1)
 
     return path_to_file
+
+
+def env_path(path_to_env):
+    if not os.path.isfile(path_to_env):
+        logger.info("Env file {} not found or not a file.".format(path_to_env))
+        return None
+
+    return path_to_env
 
 
 def github_url(url):
@@ -38,9 +46,16 @@ def github_url(url):
 def get_parser():
     parser = argparse.ArgumentParser(
                  prog="{}".format(UTIL_NAME.upper()),
+                 formatter_class=argparse.ArgumentDefaultsHelpFormatter,
                  description="Program help to find similar parts of source "
                              "codes for the different languages.",
              )
+    parser.add_argument(
+        "-env", "--environment",
+        help="Path to the environment file with GitHub access token.",
+        type=env_path,
+        default=".env",
+    )
     parser.add_argument(
         "--mode",
         help="Choose one of the following modes of searching plagiarism. "
@@ -53,7 +68,7 @@ def get_parser():
         "-t", "--threshold",
         help="Threshold of analyzer which classifies two work as same. "
              "If this number is too large, such as 99, then completely matching jobs will be found. "
-             "Otherwise, if this number is small, such as 50, then all work with minimal similarity will be found. ",
+             "Otherwise, if this number is small, such as 50, then all work with minimal similarity will be found.",
         type=int,
         default=65,
         choices=range(50, 100),
@@ -61,7 +76,7 @@ def get_parser():
     )
     parser.add_argument(
         '-v', '--version',
-        help="Print current version number and exit",
+        help="Print current version number and exit.",
         action="version",
         version="{} {}".format(UTIL_NAME, UTIL_VERSION)
     )
@@ -69,7 +84,7 @@ def get_parser():
     parser_required = parser.add_argument_group("required options")
     parser_required.add_argument(
         "-ext", "--extension",
-        help="Extension responsible for the analyzed programming language",
+        help="Extension responsible for the analyzed programming language.",
         type=str,
         choices=["py", "cpp"],
         required=True
@@ -85,26 +100,26 @@ def get_parser():
     parser_github.add_argument(
         "-e", "--regexp",
         type=str,
-        help="A regular expression to filter searching repositories on GitHub"
+        help="A regular expression to filter searching repositories on GitHub."
     )
     parser_github.add_argument(
         "-gf", "--github-files",
         metavar="GITHUB_FILE",
         type=github_url,
-        help="URL to file in a GitHub repository",
+        help="URL to file in a GitHub repository.",
         nargs="+",
         default=[]
     )
     parser_github.add_argument(
         "-g", "--github-user",
         type=str,
-        help="GitHub organisation/user name"
+        help="GitHub organisation/user name."
     )
     parser_github.add_argument(
         "-gp", "--github-project-folders",
         metavar="GITHUB_PROJECT_FOLDER",
         type=github_url,
-        help="Path to a GitHub project folder",
+        help="Path to a GitHub project folder.",
         nargs="+",
         default=[]
     )
@@ -114,7 +129,7 @@ def get_parser():
         "-f", "--files",
         metavar="FILE",
         type=file_path,
-        help="Full path to files on a computer",
+        help="Full path to files on a computer.",
         nargs="+",
         default=[]
     )
@@ -122,7 +137,7 @@ def get_parser():
         "-d", "--directories",
         metavar="DIRECTORY",
         type=dir_path,
-        help="Full path to a local directories with project/files",
+        help="Full path to a local directories with project/files.",
         nargs="+",
         default=[]
     )

--- a/src/sbin/codeplag
+++ b/src/sbin/codeplag
@@ -1,9 +1,9 @@
 #!/usr/bin/env python3
 
-import os
+import subprocess
 import sys
 
-command = 'python3 -m codeplag'
-arguments = ' '.join(sys.argv[1:])
-
-os.system('{} {}'.format(command, arguments))
+try:
+    return_code = subprocess.run(['python3', '-m', 'codeplag'] + sys.argv[1:], check=True)
+except subprocess.CalledProcessError as error:
+    sys.exit(error.returncode)

--- a/src/webparsers/github_parser.py
+++ b/src/webparsers/github_parser.py
@@ -92,13 +92,12 @@ class GitHubParser:
                 'Authorization': 'token ' + self.__access_token,
             })
 
-        # TODO: Check Ethernet connection and requests limit
-        # requests.exceptions.ConnectionError
+        # Check Ethernet connection and requests limit
         try:
             response = requests.get(address + api_url, headers=headers,
                                     params=params)
         except requests.exceptions.ConnectionError as err:
-            GitHubParser.logger.error("Connection error. Please check the Internet connection")
+            GitHubParser.logger.error("Connection error. Please check the Internet connection.")
             GitHubParser.logger.debug(str(err))
             exit(1)
 

